multi_server_environment setting now disallows config edits from Web UI

nina-py opened the pull request:
### Description:

Turning the multi_server_environment setting on now hides the following
settings from the Administration -> Settings -> General settings page:

- Archiving settings
- Email server
- Trusted Matomo hostname
- Update settings

...so that users cannot set these separately on different instances of Matomo. Assuming I got these right, are there any other settings that should be disabled when `multi_server_environment` is turned on?

Fixes #14390.

### Review

* [ ] Functional review done
* [ ] Usability review done (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] Security review done [see checklist](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] Code review done
* [ ] Tests were added if useful/possible
* [ ] Reviewed for breaking changes
* [ ] Developer changelog updated if needed
* [ ] Documentation added if needed
* [ ] Existing documentation updated if needed

nina-py committed:
multi_server_environment setting now disallows config edits from Web UI

Turning the multi_server_environment setting on now hides the following
settings from the Administration -> Settings -> General settings page:

- Archiving settings
- Email server
- Trusted Matomo hostname
- Update settings

...so that users cannot set these separately on different instances of Matomo.

Fixes #14390.

tsteur commented:
Makes sense @nina-py 

Might need to add a note in `config/global.ini.php` for `multi_server_environment` that this flag doesn't need to be enabled when the config file is on a shared filesystem like NFS or EFS.

nina-py committed:
Added a note for multi_server_environment setting about shared filesystems

nina-py commented:
Thanks @tsteur, added a note to the global config file.

nina-py commented:
Hi @tsteur , is there anything else that needs to be done in this PR? Or are you guys just way too busy at the moment following the 4.0.x release?

tsteur commented:
@nina-py at the moment we're indeed quite busy with the Matomo 4 release. @sgiehl can you do another review maybe?

sgiehl commented on plugins/CoreAdminHome/templates/generalSettings.twig:
```diff
@@ -10,7 +10,7 @@
     {{ ajax.loadingDiv() }}
 
-{% if isGeneralSettingsAdminEnabled %}
+{% if isGeneralSettingsAdminEnabled and not isMultiServerEnvironment %}
```
I guess that is actually not needed to hide the archiving settings. They are stored in the database when configured through UI. See https://github.com/matomo-org/matomo/blob/f5e9420a987340b036fa342e876ab92e314f4ec7/core/ArchiveProcessor/Rules.php#L265-L272

Only if that values is not available the one in config should be used.

Hiding the email settings only makes more sense, as they are stored in config only.

nina-py replied:
Thanks @sgiehl, I'll tackle this over the weekend.

nina-py replied:
@sgiehl, I updated the template to hide only the Email Server settings if in multi-server environment.

sgiehl commented on plugins/CoreUpdater/SystemSettings.php:
```diff
@@ -52,19 +53,20 @@ protected function init()
+                        && CoreAdminController::isGeneralSettingsAdminEnabled()
+                        && SettingsPiwik::isMultiServerEnvironment() === false;
         $this->releaseChannel = $this->createReleaseChannel();
         $this->releaseChannel->setIsWritableByCurrentUser($isWritable);
 
-        $isWritable = $isWritable && PluginUpdateCommunication::canBeEnabled();
-        $this->sendPluginUpdateEmail = $this->createSendPluginUpdateEmail();
-        $this->sendPluginUpdateEmail->setIsWritableByCurrentUser($isWritable);
-
-        $isWritable = Piwik::hasUserSuperUserAccess() && CoreAdminController::isGeneralSettingsAdminEnabled();
         $dbSettings = new Settings();
         if ($isWritable && $dbSettings->getUsedCharset() !== 'utf8mb4' && DbHelper::getDefaultCharset() === 'utf8mb4') {
             $this->updateToUtf8mb4 = $this->createUpdateToUtf8mb4();
         }
+
+        $isWritable = $isWritable && PluginUpdateCommunication::canBeEnabled();
+        $this->sendPluginUpdateEmail = $this->createSendPluginUpdateEmail();
+        $this->sendPluginUpdateEmail->setIsWritableByCurrentUser($isWritable);
```
Actually only the release channel is a configuration that is managed in config. The other values are "normal" settings, stored in database. So for those it wouldn't hurt to have them writable.

nina-py replied:
Thank you for that, I have updated the code to reflect that (and did a very minor refactor to remove duplicate code).

nina-py committed:
- multi_server_environment now only hides email server settings from the general
settings page.

- Only making the release channel not writable in CoreUpdater->SystemSettings,
and leaving the other settings as before.

nina-py commented:
Hi @sgiehl, in addition to the changes you requested earlier, I have updated a UI screenshot (an extra comment for the setting about half-way through the screenshot): https://builds-artifacts.matomo.org/matomo-org/matomo/4.x-dev/44409/UIIntegrationTest_admin_diagnostics_configfile.png. CI build in progress.

As far as I can see, the other failures in those two jobs are not related to the changes introduced in this PR. I've also rebased the branch. Please review.

nina-py commented:
@sgiehl, just to follow up on the most recent CI build for this PR - the updated screenshot matches but there is one failure elsewhere with the following:

```bash
There was 1 failure:

1) Piwik\Tests\Integration\ReleaseCheckListTest::test_screenshotsStoredInLfs

Some Screenshots are not stored in LFS: tests/UI/expected-screenshots/UIIntegrationTest_admin_diagnostics_configfile.png

Failed asserting that an array is empty.

/home/travis/build/matomo-org/matomo/tests/PHPUnit/Integration/ReleaseCheckListTest.php:180
```

Do I need to do anything about this one?

sgiehl commented:
@nina-py  We are storing our screenshots in Git LFS to not waste space in this repo. You may need to install Git LFS to update the the srceenshots properly

nina-py committed:
Update UI test

Updated UIIntegrationTest_admin_diagnostics_configfile.png - with Git LFS
this time around.

nina-py commented:
Thank you @sgiehl, I think I got this right this time around - the commit diff shows

```
Git LFS file not shown 
```

but the build is yet to run.

sgiehl reviewed (approved):
Looks good now 👍  @nina-py

nina-py commented:
Thank you @sgiehl for all your help with this one!
